crypto/elliptic-curve.ts

```diff
@@ -703,3 +716,369 @@ function createCurveAffine({
+): GroupProjective {
+  let { x: X1, y: Y1, z: Z1 } = g;
+  let { x: X2, y: Y2, z: Z2 } = h;
+
+  // A = Z1 * Z2
+  let A = mod(Z1 * Z2, p);
+  // B = A^2
+  let B = mod(A * A, p);
+  // C = X1 * X2
+  let C = mod(X1 * X2, p);
+  // D = Y1 * Y2
+  let D = mod(Y1 * Y2, p);
+  // E = d * C * D
+  let E = mod(d * C * D, p);
+  // F = B - E
+  let F = mod(B - E, p);
+  // G = B + E
+  let G = mod(B + E, p);
+
+  return {
+    x: mod(A * F * ((X1 + Y1) * (X2 + Y2) - C - D), p),
+    y: mod(A * G * (D - a * C), p),
+    z: mod(F * G, p),
+  };
+}
+
+// https://www.hyperelliptic.org/EFD/g1p/auto-twisted-projective.html
+// https://eprint.iacr.org/2008/013.pdf Section 6
+function projectiveDoubleTwisted(
+  g: GroupProjective,
+  p: bigint,
+  a: bigint
+): GroupProjective {
+  let { x: X, y: Y, z: Z } = g;
+
+  // B = (X + Y)^2
+  let B = mod((X + Y) ** 2n, p);
+  // C = X^2
+  let C = mod(X * X, p);
+  // D = Y^2
+  let D = mod(Y * Y, p);
+  // E = a * C
+  let E = mod(a * C, p);
+  // F = E + D
+  let F = mod(E + D, p);
+  // H = Z^2
+  let H = mod(Z * Z, p);
+  // J =  F - 2 * H
+  let J = mod(F - 2n * H, p);
+
+  return {
+    x: mod((B - C - D) * J, p),
+    y: mod(F * (E - D), p),
+    z: mod(F * J, p),
+  };
+}
+
+function projectiveScaleTwisted(
+  g: GroupProjective,
+  x: bigint | boolean[],
+  p: bigint,
+  a: bigint,
+  d: bigint
+) {
+  let bits = typeof x === 'bigint' ? bigIntToBits(x) : x;
+  let h = projectiveZeroTwisted;
+  for (let bit of bits) {
+    if (bit) h = projectiveAddTwisted(h, g, p, a, d);
+    g = projectiveDoubleTwisted(g, p, a);
+  }
+  return h;
+}
+
+function projectiveFromTwisted({
+  x,
+  y,
+  infinity,
+}: GroupTwisted): GroupProjective {
+  if (infinity) return projectiveZeroTwisted;
+  return { x, y, z: 1n };
+}
+
+// The twisted curve with equation
+// a * x^2 + y^2 = 1 + d * x^2 * y^2
+// in projective coordinates is represented as
+// a * X^2 * Z^2 + Y^2 Z^2 = Z^4 + d * X^2 * Y^2
+// where x = X/Z, y = Y/Z, and Z ≠ 0
+function projectiveToTwisted(g: GroupProjective, p: bigint): GroupTwisted {
+  let z = g.z;
+  if (z === 0n) {
+    // degenerate case
+    return twistedZero;
+  } else if (z === 1n && g.x === 0n && g.y === 1n) {
+    // special case for the zero point
+    return twistedZero;
+  } else if (z === 1n) {
+    // any other normalized affine form
+    return { x: g.x, y: g.y, infinity: false };
+  } else {
+    let zinv = inverse(z, p)!; // we checked for z === 0, so inverse exists
+    // x/z
+    let x = mod(g.x * zinv, p);
+    // y/z
+    let y = mod(g.y * zinv, p);
+    return { x: x, y: y, infinity: false };
+  }
+}
+
+type CurveTwisted = ReturnType<typeof createCurveTwisted>;
+
+// Creates twisted Edwards curves in the form
+// a * x^2 + y^2 = 1 + d * x^2 * y^2
+// with a ≠ 0, d ≠ 0 and a ≠ d
+function createCurveTwisted({
+  name,
+  modulus: p,
+  order,
+  cofactor,
+  generator,
+  a,
+  d,
+}: TwistedCurveParams) {
+  let hasCofactor = cofactor !== undefined && cofactor !== 1n;
```

Comment:
Oh boy. Don't tell me that you have to check both equalities in Typescript?!?!?!? :scream: Is it not `cofactor !== 1n` sufficient?

Reply:
Ah no, ok, it is because you accept it to be undefined for a cofactor of 1. Closing.